### ISSUE TITLE
drawbacks of formula of calculation euclidean similarity

### ISSUE THREAD

**Issue by dtanalytic** [langchain-ai/langchain#32058]
### Checked other resources

- [x] This is a bug, not a usage question. For questions, please use the LangChain Forum (https://forum.langchain.com/).
- [x] I added a clear and descriptive title that summarizes this issue.
- [x] I used the GitHub search to find a similar question and didn't find it.
- [x] I am sure that this is a bug in LangChain rather than my code.
- [x] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).
- [x] I read what a minimal reproducible example is (https://stackoverflow.com/help/minimal-reproducible-example).
- [x] I posted a self-contained, minimal, reproducible example. A maintainer can copy it and run it AS IS.

### Example Code

!pip install -q faiss-cpu
!pip install -q langchain-huggingface
!pip install langchain-community -q

``` python
from langchain_community.vectorstores import FAISS
from langchain_core.documents import Document
from langchain_community.vectorstores.faiss import DistanceStrategy
import numpy as np

from langchain_huggingface import HuggingFaceEmbeddings
import torch

device = 'cuda' if torch.cuda.is_available() else 'cpu'


mn = 'basel/ATTACK-BERT'
query = " execute malware loaders"
embed_wrapper = HuggingFaceEmbeddings(model_name=mn, 
                                   model_kwargs={'device': device})


sents = ['Sidewinder has used JavaScript to drop and execute malware loaders.',
 'Sidewinder has used PowerShell to drop and execute malware loaders.',
 'It includes a module on internet threats designed to help end users learn how to identify and protect themselves from various types of phishing attacks.',
 'It includes a module on Internet threats designed to help end users learn how to identify and protect themselves from various types of phishing attacks.',
 'regexp_url (accessed Apr. 25, 2023).',
 'regexp_url (accessed Apr. 28, 2023).',
 'It’s unclear whether Victim 1 was impacted by Trigona.',
 'It’s unclear whether Victim 2 was impacted by Trigona.',
 'Use a reputed anti-virus and internet security software package on your connected devices, including PC, laptop, and mobile.',
 'Use a reputed anti-virus and Internet security software package on your connected devices, including PC, laptop, and mobile.']


N = 100
db = FAISS.from_texts(sents[:N], embed_wrapper, ids=range(len(sents[:N])), 
                      distance_strategy = DistanceStrategy.EUCLIDEAN_DISTANCE, normalize_L2=True)

docs_scores = db.similarity_search_with_relevance_scores(query, k=4,
                                                          score_threshold = 0)
for doc, score in docs_scores:
    print(doc.page_content, doc.metadata, score)



embedding = db._embed_query(query)


print(db.index.search(np.array([embedding]), 4))
print('db.index.search return the square of euclidean distance. if you take the first doc:')

embedding1 = db._embed_query(db.get_by_ids([1])[0].page_content)

print(f'The square of euclidean distance {np.linalg.norm(np.array(embedding1)-np.array(embedding))**2}')
print(f'Then similarity according to formula: {1-0.6841346 /(2**0.5)}')

db = FAISS.from_texts(sents[:N], embed_wrapper,
                      distance_strategy = DistanceStrategy.COSINE, normalize_L2=True)

embedding = db._embed_query(query)

print(f'if you change distance_strategy to DistanceStrategy.COSINE, you\'ll get the same distances as in euclidean case')
print(db.index.search(np.array([embedding]), 4))
```

### Error Message and Stack Trace (if applicable)

!pip install -q faiss-cpu
!pip install -q langchain-huggingface
!pip install langchain-community -q

### Description

As far as I understand sqrt of 2 is used in fraction of similarity measure, because of formula below with max distance when vectors are positive, but

My points are:
1) we can have negative vectors too, than max distance is 2
2) in the numerator is used a square of euclidean distance than in denominator must be too
3) DistanceStrategy.COSINE uses the same distance calcs as euclidean case, why
4) why is it not recommended to normalize in case  of not euclidean distance...
Normalizing L2 is not applicable for metric type: DistanceStrategy.COSINE

<img width="770" height="282" alt="Image" src="https://github.com/user-attachments/assets/ec8065bd-3f54-4f0f-8ffa-e1dddd8b2403" />

### System Info

System Information
------------------
> OS:  Linux
> OS Version:  #1 SMP PREEMPT_DYNAMIC Sun Mar 30 16:01:29 UTC 2025
> Python Version:  3.11.13 (main, Jun  4 2025, 08:57:29) [GCC 11.4.0]

Package Information
-------------------
> langchain_core: 0.3.68
> langchain: 0.3.26
> langchain_community: 0.3.27
> langsmith: 0.4.4
> langchain_huggingface: 0.3.0
> langchain_text_splitters: 0.3.8

Optional packages not installed
-------------------------------
> langserve

Other Dependencies
------------------
> aiohttp<4.0.0,>=3.8.3: Installed. No version info available.
> async-timeout<5.0.0,>=4.0.0;: Installed. No version info available.
> dataclasses-json<0.7,>=0.5.7: Installed. No version info available.
> httpx: 0.28.1
> httpx-sse<1.0.0,>=0.4.0: Installed. No version info available.
> huggingface-hub>=0.30.2: Installed. No version info available.
> jsonpatch<2.0,>=1.33: Installed. No version info available.
> langchain-anthropic;: Installed. No version info available.
> langchain-aws;: Installed. No version info available.
> langchain-azure-ai;: Installed. No version info available.
> langchain-cohere;: Installed. No version info available.
> langchain-community;: Installed. No version info available.
> langchain-core<1.0.0,>=0.3.51: Installed. No version info available.
> langchain-core<1.0.0,>=0.3.65: Installed. No version info available.
> langchain-core<1.0.0,>=0.3.66: Installed. No version info available.
> langchain-deepseek;: Installed. No version info available.
> langchain-fireworks;: Installed. No version info available.
> langchain-google-genai;: Installed. No version info available.
> langchain-google-vertexai;: Installed. No version info available.
> langchain-groq;: Installed. No version info available.
> langchain-huggingface;: Installed. No version info available.
> langchain-mistralai;: Installed. No version info available.
> langchain-ollama;: Installed. No version info available.
> langchain-openai;: Installed. No version info available.
> langchain-perplexity;: Installed. No version info available.
> langchain-text-splitters<1.0.0,>=0.3.8: Installed. No version info available.
> langchain-together;: Installed. No version info available.
> langchain-xai;: Installed. No version info available.
> langchain<1.0.0,>=0.3.26: Installed. No version info available.
> langsmith-pyo3: Installed. No version info available.
> langsmith>=0.1.125: Installed. No version info available.
> langsmith>=0.1.17: Installed. No version info available.
> langsmith>=0.3.45: Installed. No version info available.
> numpy>=1.26.2;: Installed. No version info available.
> numpy>=2.1.0;: Installed. No version info available.
> openai-agents: Installed. No version info available.
> opentelemetry-api: Installed. No version info available.
> opentelemetry-exporter-otlp-proto-http: Installed. No version info available.
> opentelemetry-sdk: Installed. No version info available.
> orjson: 3.10.18
> packaging: 24.2
> packaging<25,>=23.2: Installed. No version info available.
> pydantic: 2.11.7
> pydantic-settings<3.0.0,>=2.4.0: Installed. No version info available.
> pydantic<3.0.0,>=2.7.4: Installed. No version info available.
> pydantic>=2.7.4: Installed. No version info available.
> pytest: 8.3.5
> PyYAML>=5.3: Installed. No version info available.
> requests: 2.32.3
> requests-toolbelt: 1.0.0
> requests<3,>=2: Installed. No version info available.
> rich: 13.9.4
> sentence-transformers>=2.6.0;: Installed. No version info available.
> SQLAlchemy<3,>=1.4: Installed. No version info available.
> tenacity!=8.4.0,<10,>=8.1.0: Installed. No version info available.
> tenacity!=8.4.0,<10.0.0,>=8.1.0: Installed. No version info available.
> tokenizers>=0.19.1: Installed. No version info available.
> transformers>=4.39.0;: Installed. No version info available.
> typing-extensions>=4.7: Installed. No version info available.
> zstandard: 0.23.0

**Comment by CuberMessenger** [comment]
@dtanalytic Thank you for the demonstration—I now see the issue.
I’ll provide a clear example to reproduce the problem and resolve it shortly.

@eugeneo17 Thank you for your attention; I’ll take care of it.

**Comment by CuberMessenger** [comment]
### Summary of the bug
In [langchain_community.vectorstores.FAISS.similarity_search_with_score_by_vector](https://github.com/langchain-ai/langchain-community/blob/0544f051b0c7d59ee364e5402dec99c177aa2b12/libs/community/langchain_community/vectorstores/faiss.py#L387), the function should returns **L2 distance** as the comment mentioned. But the function returns **square of L2 distance**.

It's because the internal FAISS search function at https://github.com/langchain-ai/langchain-community/blob/0544f051b0c7d59ee364e5402dec99c177aa2b12/libs/community/langchain_community/vectorstores/faiss.py#L417 actually returns the square of L2 distance.

Also, I don't think the FAISS in `langchain_community.vectorstores.FAISS` currently supports cosine similarity if you check https://github.com/langchain-ai/langchain-community/blob/0544f051b0c7d59ee364e5402dec99c177aa2b12/libs/community/langchain_community/vectorstores/faiss.py#L997-L1001. @dtanalytic 

I will address it soon. Just need to make sure the other supported distance strategy will be fine.
I'm already working on this. @eugeneo17 Thanks.

### To demonstrate the problem, let's be clear about `distance` and `similarity`.

Consider two vectors $a, b \in ℝ^n$.  
In the Euclidean case (or L2 case), `distance` means

$$
d_{euc} = \sqrt{\sum_{i=1}^n (a_i - b_i)^2}.
$$

As long as the two vectors are normalized to the unit hypersphere of dimension $n$, $d_{euc} \in [0, 2]$, which is true if you see
https://github.com/langchain-ai/langchain-community/blob/0544f051b0c7d59ee364e5402dec99c177aa2b12/libs/community/langchain_community/vectorstores/faiss.py#L415-L416.

Then it naturally defines a Euclidean similarity as

$$
s_{euc} = 1 - \frac{d_{euc}}{\sqrt{2}},
$$

$$
s_{euc} \in [0, 1].
$$

### A minimal example
Here's a minimal example of the bug. Also, a patched version to show the fix.

```python
import numpy as np
from langchain_core.documents import Document
from langchain_community.vectorstores import FAISS
from langchain_community.vectorstores.faiss import DistanceStrategy
from langchain_google_genai import GoogleGenerativeAIEmbeddings

import getpass


def get_google_embedding_model(name="models/text-embedding-004"):
    api_key = getpass.getpass("Enter your Google API Key: ")

    embedding_model = GoogleGenerativeAIEmbeddings(
        model=name,
        google_api_key=api_key,
        transport="rest",
        timeout=30,
        max_retries=3,
    )

    return embedding_model


embedding_model = get_google_embedding_model()

sents = [
    "Sidewinder has used JavaScript to drop and execute malware loaders.",
]

db_euc = FAISS.from_texts(
    sents,
    embedding_model,
    ids=range(len(sents)),
    distance_strategy=DistanceStrategy.EUCLIDEAN_DISTANCE,
    normalize_L2=True,
)

query = "execute malware loaders"

embedding_doc = np.array(db_euc._embed_query(db_euc.get_by_ids([0])[0].page_content))
embedding_euc = np.array(db_euc._embed_query(query))

# Euclidean distance and similarity by hand
euc_distence = np.sqrt(((embedding_doc - embedding_euc) ** 2).sum())
euc_similarity = 1 - euc_distence / (2**0.5)

print(f"[By hand] Euclidean distance: {euc_distence}")
print(f"[By hand] Euclidean similarity: {euc_similarity}")
"""
[By hand] Euclidean distance: 0.6381204577868655
[By hand] Euclidean similarity: 0.5487806970850433
"""

# Score by FAISS
score = db_euc.similarity_search_with_score(query, k=1)[0][1]

print(f"[FAISS] Score: {score}")
"""
[FAISS] Score: 0.40719807147979736

The score is actually the square of the euclidean distance.
(0.6381204577868655) ** 2) = 0.4071977186461187
"""

# Relevance score by FAISS
relevance_score = db_euc.similarity_search_with_relevance_scores(query, k=1)[0][1]
print(f"[FAISS] Relevance score: {relevance_score}")
"""
[FAISS] Relevance score: 0.7120674848556519

This relevance score is actually
0.7120674848556519 = 1 - (0.40719807147979736 / (2**0.5))
"""


### Monkey patch the fix ###
from typing import Any, Callable, Dict, List, Optional, Union, Tuple
import operator
from langchain_community.vectorstores.faiss import dependable_faiss_import


def similarity_search_with_score_by_vector(
    self,
    embedding: List[float],
    k: int = 4,
    filter: Optional[Union[Callable, Dict[str, Any]]] = None,
    fetch_k: int = 20,
    **kwargs: Any,
) -> List[Tuple[Document, float]]:
    """Return docs most similar to query.

    Args:
        embedding: Embedding vector to look up documents similar to.
        k: Number of Documents to return. Defaults to 4.
        filter (Optional[Union[Callable, Dict[str, Any]]]): Filter by metadata.
            Defaults to None. If a callable, it must take as input the
            metadata dict of Document and return a bool.
        fetch_k: (Optional[int]) Number of Documents to fetch before filtering.
                  Defaults to 20.
        **kwargs: kwargs to be passed to similarity search. Can include:
            score_threshold: Optional, a floating point value between 0 to 1 to
                filter the resulting set of retrieved docs

    Returns:
        List of documents most similar to the query text and L2 distance
        in float for each. Lower score represents more similarity.
    """
    faiss = dependable_faiss_import()
    vector = np.array([embedding], dtype=np.float32)
    if self._normalize_L2:
        faiss.normalize_L2(vector)
    scores, indices = self.index.search(vector, k if filter is None else fetch_k)
    scores = np.sqrt(scores) ############################ ADDED ###########################################
    docs = []

    if filter is not None:
        filter_func = self._create_filter_func(filter)

    for j, i in enumerate(indices[0]):
        if i == -1:
            # This happens when not enough docs are returned.
            continue
        _id = self.index_to_docstore_id[i]
        doc = self.docstore.search(_id)
        if not isinstance(doc, Document):
            raise ValueError(f"Could not find document for id {_id}, got {doc}")
        if filter is not None:
            if filter_func(doc.metadata):
                docs.append((doc, scores[0][j]))
        else:
            docs.append((doc, scores[0][j]))

    score_threshold = kwargs.get("score_threshold")
    if score_threshold is not None:
        cmp = (
            operator.ge
            if self.distance_strategy
            in (DistanceStrategy.MAX_INNER_PRODUCT, DistanceStrategy.JACCARD)
            else operator.le
        )
        docs = [
            (doc, similarity)
            for doc, similarity in docs
            if cmp(similarity, score_threshold)
        ]
    return docs[:k]


FAISS.similarity_search_with_score_by_vector = similarity_search_with_score_by_vector

### Monkey patch the fix ###

# Test the fixed score by FAISS
score = db_euc.similarity_search_with_score(query, k=1)[0][1]

print(f"[Fixed FAISS] Score: {score}")
"""
[FAISS] Score: 0.638120710849762
This matches the by-hand calculation of the Euclidean distance.
"""

# Test the fixed relevance score by FAISS
score = db_euc.similarity_search_with_relevance_scores(query, k=1)[0][1]
print(f"[Fixed FAISS] Relevance score: {score}")
"""
[FAISS] Relevance score: 0.5487805008888245
This matches the by-hand calculation of the Euclidean similarity.
"""
```
